mcp-server-open-api-spec/CLAUDE.MD

```diff
@@ -0,0 +1,371 @@
+
+### Pydantic Schema Validation
+All MCP tool arguments and responses flow through Pydantic models:
+```python
+# In server.py - convert to dict for MCP
+return result.model_dump()
+
+# In handlers.py - receive validated schema
+def list_api_endpoints(args: ListApiEndpointsRequest) -> ListApiEndpointsResponse:
+    ...
+```
+
+### Context Variable for Headers
+Authentication headers stored in `contextvars.ContextVar` for thread-safety:
+```python
+# Middleware sets
+header_store.set(dict(request.headers))
+
+# Utils reads
+headers = header_store.get({})
+token = headers.get("x-forwarded-access-token")
+```
+
+### HTTP Method Mapping
+Databricks SDK uses enums, not strings:
+```python
+method_mapping = {
+    "GET": ExternalFunctionRequestHttpMethod.GET,
+    "POST": ExternalFunctionRequestHttpMethod.POST,
+    # ...
+}
+```
+
+### Parameter Handling by Method
+- **GET/DELETE**: Parameters converted to query string
+- **POST/PUT/PATCH**: Parameters sent as JSON body
+- Supports both `Dict[str, Any]` and string (with JSON parsing)
+
+### Error Handling
+- All handlers raise exceptions on error
+- `server.py` catches exceptions and returns structured error responses
+- Includes helpful error messages for debugging
+
+## Common Tasks
+
+### Adding a New MCP Tool
+1. Define Pydantic schemas in `schemas.py`
+2. Implement handler logic in `handlers.py`
+3. Register tool in `server.py` with `@mcp.tool()` decorator
+4. Add docstring for auto-generated tool description
+
+### Modifying OpenAPI Spec Loading
+- Edit `utils.py` `load_openapi_spec()` function
+- Currently loads from `{SPEC_VOLUME_PATH}/spec.json`
+- Global cache prevents repeated file reads
+
+### Changing Authentication Method
+- Modify `get_workspace_client()` in `utils.py`
+- Update middleware in `server.py` if header name changes
+- Ensure UC external connection type matches
+
+### Updating API Response Format
+- Modify `InvokeApiEndpointResponse` schema in `schemas.py`
+- Update response construction in `handlers.py` `invoke_api_endpoint()`
+- Adjust serialization logic in `server.py` if needed
+
+## Deployment
+
+### Local Development
+```bash
+uv sync                    # Install dependencies
+uv run custom-server       # Start with hot reload
+```
+
+Access at: `http://localhost:8000`
+
+### Databricks Apps Deployment
+1. Update `UC_CONNECTION_NAME` and `SPEC_VOLUME_PATH` in `app.yaml`
+2. Deploy using Databricks CLI or UI
+3. App runs command: `uv run custom-server`
+4. Access at: `https://your-app-url.databricksapps.com/mcp`
+
+### Build Hook
+`hooks/apps_build.py` provides custom build logic for Databricks Apps (exact behavior depends on implementation).
+
+## Testing & Debugging
+
+### Local Testing
+- Use MCP inspector or direct HTTP requests
+- Set logging level to `DEBUG` in `server.py` for verbose output
+- Check `/health` endpoint for basic connectivity
+
+### Debugging Tips
+1. **Auth issues**: Check `x-forwarded-access-token` header presence
```

Comment:
Can we share that if the header in not present, that means the exact scope for OBO are not present?